### PR TITLE
UCP/TEST: Support multiple endpoints between same pair of workers

### PR DESCRIPTION
**Pull request by yosefe**

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6769/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6770/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6772/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6788/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6789/ for details (Mellanox internal link).

File: src/ucp/wireup/ep_match.c [hunk 1]
```diff
@@ -102,8 +102,8 @@ static void ucp_ep_match_insert_common(ucp_ep_match_ctx_t *match_ctx,
 {
     ucs_assert(!(ep->flags & UCP_EP_FLAG_ON_HASH));
     ucp_ep_match_list_add_tail(list, &ucp_ep_ext_gen(ep)->ep_match.list);
-    ep->flags                     |= UCP_EP_FLAG_ON_HASH;
-    ucp_ep_ext_gen(ep)->dest_uuid  = dest_uuid;
+    ep->flags                              |= UCP_EP_FLAG_ON_HASH;
```
Comment (evgeny-leksikov) [inline src/ucp/wireup/ep_match.c]: Maybe rename this flag to smt like `UCP_EP_FLAG_ON_MATCH_CTX`? we don't have direct worker level hash storage anymore. It's encapsulated inside matching context.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6797/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6800/ for details (Mellanox internal link).

File: src/ucp/wireup/wireup.c [hunk 2]
```diff
@@ -175,60 +192,71 @@ static void ucp_wireup_remote_connected(ucp_ep_h ep)
...
-    } else {
-        /* Reply for a client-server connection (client side) */
-        ep = ucp_worker_ep_find(worker, msg->ep_uuid);
-        if (ep == NULL) {
-            ucs_trace("got connection request with invalid ep_uuid 0x%"PRIx64,
-                      msg->ep_uuid);
+
+        ucp_ep_update_dest_ep_ptr(ep, msg->src_ep_ptr);
+
+        if ((ep->flags & UCP_EP_FLAG_CONNECT_REQ_QUEUED) && (remote_uuid > worker->uuid)) {
```
Comment (alinask) [inline src/ucp/wireup/wireup.c]: please document the condition

File: src/ucp/wireup/wireup.c [hunk 3]
```diff
@@ -175,60 +192,71 @@ static void ucp_wireup_remote_connected(ucp_ep_h ep)
...
         }
 
-        /* Reinsert to hash table with destination worker uuid */
-        ucs_assert(!(ep->flags & UCP_EP_FLAG_DEST_UUID_PEER));
-        ucp_ep_delete_from_hash(ep);
-        ucp_ep_ext_gen(ep)->dest_uuid = remote_address->uuid;
-        ep->flags                    |= UCP_EP_FLAG_DEST_UUID_PEER;
-        ucp_ep_add_to_hash(ep);
-
         ep_init_flags = UCP_EP_CREATE_AM_LANE;
```
Comment (alinask) [inline src/ucp/wireup/wireup.c]: please remove the flag

File: src/ucp/core/ucp_ep.inl [hunk 4]
```diff
@@ -132,16 +132,42 @@ static inline ucp_ep_h ucp_ep_from_ext_proto(ucp_ep_ext_proto_t *ep_ext)
...
-        ucs_assert(ep->flags & UCP_EP_FLAG_DEST_UUID_PEER);
-        ucp_wireup_send_request(ep, ucp_ep_ext_gen(ep)->dest_uuid);
+    if (ep->flags & UCP_EP_FLAG_DEST_EP) {
+        return UCS_OK;
     }
+
+    return ucp_wireup_connect_remote(ep, lane);
+}
+
+static inline void ucp_ep_update_dest_ep_ptr(ucp_ep_h ep, uintptr_t ep_ptr)
```
[the remark this answers is not in the document]
Comment (yosefe) [reply src/ucp/core/ucp_ep.inl]: it's not fast path

[the remark this answers is not in the document]
Comment (brminich) [reply src/ucp/core/ucp_ep.inl]: ok

File: src/ucp/core/ucp_ep.c [hunk 5]
```diff
@@ -449,7 +404,7 @@ ucs_status_t ucp_ep_create(ucp_worker_h worker, const ucp_ep_params_t *params,
 {
     ucs_status_t status;
     unsigned flags;
-    ucp_ep_h ep;
+    ucp_ep_h ep = NULL;
```
[the remark this answers is not in the document]
Comment (brminich) [reply src/ucp/core/ucp_ep.c]: Checked that if ucp_ep_create_to_sock_addr() or ucp_ep_create_to_worker_addr() would return an error, `*ep_p` will not be assigned to `ep`. So the assignment looks excessive.

[the remark this answers is not in the document]
Comment (brminich) [reply src/ucp/core/ucp_ep.c]: Ok then. I noticed this assignment, because it is not in the top of the function :)

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6805/ for details (Mellanox internal link).

**Pull request merged**